contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,288 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {IERC5267} from "../../interfaces/IERC5267.sol";
+import {MessageHashUtils} from "./MessageHashUtils.sol";
+import {Bytes} from "../Bytes.sol";
+
+/**
+ * @dev Utilities to process https://eips.ethereum.org/EIPS/eip-7739[ERC-7739] typed data signatures
+ * that are specific to an {EIP712} domain.
+ *
+ * This library provides methods to wrap, unwrap and operate over typed data signatures with a defensive
+ * rehashing mechanism that includes the application's {EIP712-_domainSeparatorV4} and preserves
+ * readability of the signed content using an EIP-712 nested approach.
+ *
+ * A smart contract domain can validate a signature for a typed data structure in two ways:
+ *
+ * - As an application validating a typed data signature. See {toNestedTypedDataHash}.
+ * - As a smart contract validating a raw message signature. See {toNestedPersonalSignHash}.
```

Comment:
What is the difference between an application and a smart contract? I don't understand this comment.

Reply:
should be that an application is third-party to the smart contract.

the point here is that a personal signature is "personal" for the smart contract (its own domain), and an application is an external domain.

not sure if it makes sense, though

Reply:
Personal signatures can also be requested and validated by third-party contracts. The distinction is not about the validating domain but the format of the signature.